### PR TITLE
Feat/os.network interfaces

### LINKED ISSUE #1550: os.networkInterfaces and os.cpus return empty objects
### What version of Bun is running?

0.3.0

### What platform is your computer?

Darwin 20.6.0 Darwin Kernel Version 20.6.0: Tue Jun 21 20:50:28 PDT 2022; root:xnu-7195.141.32~1/RELEASE_X86_64 x86_64 i386

### What steps can reproduce the bug?

```ts
import { cpus, networkInterfaces } from 'os';

console.log({
  cpus: cpus(),
  networkInterfaces: networkInterfaces(),
});
```

### How often does it reproduce? Is there a required condition?

100% of the time, tested on both macos and linux.

### What is the expected behavior?

```
{
  cpus: [
    {
      model: 'Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz',
      speed: 2500,
      times: [Object]
    },
    {
      model: 'Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz',
      speed: 2500,
      times: [Object]
    },
    {
      model: 'Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz',
      speed: 2500,
      times: [Object]
    },
    {
      model: 'Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz',
      speed: 2500,
      times: [Object]
    },
    {
      model: 'Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz',
      speed: 2500,
      times: [Object]
    },
    {
      model: 'Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz',
      speed: 2500,
      times: [Object]
    },
    {
      model: 'Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz',
      speed: 2500,
      times: [Object]
    },
    {
      model: 'Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz',
      speed: 2500,
      times: [Object]
    }
  ],
  networkInterfaces: {
    lo0: [ [Object], [Object], [Object] ],
    en0: [ [Object], [Object] ],
    awdl0: [ [Object] ],
    llw0: [ [Object] ],
    utun0: [ [Object] ],
    utun1: [ [Object] ],
    utun2: [ [Object] ],
    utun3: [ [Object] ],
    utun4: [ [Object] ],
    utun5: [ [Object] ]
  }
}
```

### What do you see instead?

```
{ cpus: [], networkInterfaces: { } }
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by jwhear**
Implements `os.networkInterfaces` for both Linux and OSX, ~~partially resolving #1550~~ fully resolves #1550 now that  #2115 is merged.

Testing locally on Linux produces exact same results as Node aside from field ordering.
Tested on x86_64 OSX, same results as Node.

### CODE REVIEW

**Comment by ThatOneBro** [comment]
Output seems to be the same as Node on OSX12 x86_64. Looks good to me
Screen Shot 2023-02-24 at 2 51 31 PM

File: src/bun.js/node/node_os.zig [hunk 1]
```diff
@@ -294,10 +294,193 @@ pub const Os = struct {
...
+            );
+
+            // mac <string> The MAC address of the network interface
+            {
+                // We need to search for the link-layer interface whose name matches this one
+                var ll_it = interface_start;
+                const maybe_ll_addr = while (ll_it) |ll_iface| : (ll_it = ll_iface.ifa_next) {
+                    if (helpers.skip(ll_iface) or !helpers.isLinkLayer(ll_iface)) continue;
+
+                    const ll_name = std.mem.sliceTo(ll_iface.ifa_name, 0);
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_os.zig]: `bun.sliceTo` uses `strlen` which is much faster than std.mem.len

```suggestion
                    const ll_name = bun.sliceTo(ll_iface.ifa_name, 0);
```

File: src/bun.js/node/node_os.zig [hunk 2]
```diff
@@ -294,10 +294,193 @@ pub const Os = struct {
...
+                    if (!std.mem.startsWith(u8, ll_name, interface_name)) continue;
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_os.zig]:
```suggestion
                    if (!strings.hasPrefix(ll_name, interface_name)) continue;
```

we should also make our startsWith wrapper use our faster string equality checker

image

File: src/bun.js/node/node_os.zig [hunk 3]
```diff
@@ -473,3 +656,76 @@ pub const Os = struct {
         return JSC.ZigString.static(comptime getMachineName()).toValue(globalThis);
     }
 };
+
+fn formatAddress(address: std.net.Address, into: []u8) ![]u8 {
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_os.zig:660]: nice

File: src/bun.js/node/node_os.zig [hunk 4]
```diff
@@ -294,10 +294,193 @@ pub const Os = struct {
...
+
+            globalThis.vm().throwError(globalThis, err.toErrorInstance(globalThis));
+            return JSC.JSValue.jsUndefined();
+        }
+        defer C.freeifaddrs(interface_start);
+
+
+        const helpers = struct {
+            // We'll skip interfaces that aren't actually available
+            pub fn skip(iface: *C.ifaddrs) bool {
```
Comment (Jarred-Sumner) [inline src/bun.js/node/node_os.zig:316]: nice

**Comment by Jarred-Sumner** [comment]
Thank you

**Pull request merged**